Fix git-credential helper to respect username parameter

cavanaug opened the pull request:
## Summary

- Fix `gh auth git-credential` to properly look up tokens for specific users when git provides a username parameter
- This enables multi-account authentication workflows where different repositories need different user credentials
- This addresses issue #11938 

## Changes

Previously, the git-credential helper ignored usernames passed by git (e.g., `username=dilbertwork_acme`) and always returned the active user's token. This behavior was inconsistent with `gh auth token --user USERNAME`, which correctly looks up tokens for specific users.

This PR restructures the token lookup logic to:
1. Check environment variables first (highest priority, overrides username)
2. If username provided, look up that specific user's token via `TokenForUser()`
3. Fall back to active user's token if user-specific token not found
4. Remove restrictive validation that rejected non-active usernames

## Testing

- All existing tests pass (14 git-credential tests)
- Added 3 new test cases covering user-specific token lookups
- Verified with real-world multi-user scenarios

Fixes authentication for multi-account setups where git passes user-specific credentials to the helper.

Fixes #11938

Commit:
fix(auth): git-credential helper now respects username parameter

cliAutomation commented:
Hi! Thanks for the pull request. Please ensure that this change is linked to an issue by mentioning an issue number in the description of the pull request. If this pull request would close the issue, please put the word 'Fixes' before the issue number somewhere in the pull request body. If this is a tiny change like fixing a typo, feel free to ignore this message.

Commit:
Refactor gist hostname fallback logic to avoid variable mutation

Replace buggy implementation that mutates lookupHost with cleaner helper
function approach. This fixes two issues identified in code review:

1. Lines 137-139 checked wants["host"] but should check lookupHost after
   gist prefix was stripped on line 122
2. Lines 146-153 used mutated lookupHost in fallback logic, potentially
   causing incorrect lookups

The new tryBothHosts helper function encapsulates gist hostname fallback
without state mutation, making the code clearer and less error-prone.

Commit:
Address Copilot review feedback on tryBothHosts helper

- Return fallback error when both host lookups fail (instead of original error)
- Extract activeTokenLookup helper to eliminate duplicate lambda functions
- Add comment documenting intentional error discard in fallback logic

cavanaug commented on pkg/cmd/auth/gitcredential/helper.go:
```diff
@@ -112,29 +113,86 @@ func helperRun(opts *CredentialOptions) error {
+			if fallbackErr == nil && fallbackToken != "" {
+				return fallbackToken, fallbackSrc, nil
+			}
+			return fallbackToken, fallbackSrc, fallbackErr
```
I dunno, I think this is debateable...

babakks commented:
Closing as explained in https://github.com/cli/cli/issues/11938#issuecomment-3432052977.
